libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Trace.hs

```diff
@@ -0,0 +1,283 @@
+  Reflect era =>
+  Proof era ->
+  Model era ->
+  Word64 ->
+  Word64 ->
+  [(Word64, Core.Tx era)] ->
+  GenRS era (Vector [(Word64, Core.Tx era)])
+genRsTxSeq _ _ this lastN ans | this > lastN = chop (reverse ans) []
+genRsTxSeq proof model0 this lastN ans = do
+  (tx, model1) <- genRsTxAndModel proof model0 this
+  genRsTxSeq proof model1 (this + 1) lastN ((this, tx) : ans)
+
+-- | Chop a list into random size blocks
+chop :: [(Word64, Core.Tx era)] -> [[(Word64, Core.Tx era)]] -> GenRS era (Vector [(Word64, Core.Tx era)])
+chop [] ans = pure $ Vector.fromList (reverse ans)
+chop xs ans | length xs <= 3 = pure $ Vector.fromList (reverse (xs : ans))
+chop pairs ans = do
+  maxBlockSize <- getBlocksizeMax <$> get
+  n <- lift $ choose (2 :: Int, fromIntegral maxBlockSize)
+  let block = Prelude.take n pairs
+  chop (Prelude.drop n pairs) (block : ans)
+
+-- | Generate a Vector of Blocks, and an initial LedgerState
+genTxSeq ::
+  forall era.
+  Reflect era =>
+  Proof era ->
+  GenSize ->
+  Word64 ->
+  Gen (Vector [(Word64, Core.Tx era)], GenState era)
+genTxSeq proof gensize numTx = do
+  runGenRS proof gensize (genRsTxSeq proof mNewEpochStateZero 0 numTx [])
+
+run :: IO ()
+run = do
+  (vs, _) <- generate $ genTxSeq (Babbage Mock) def 5
+  putStrLn (show (ppList (ppList (ppPair ppWord64 (pcTx (Babbage Mock)))) (Vector.toList vs)))
+
+-- ==================================================================
+-- Constructing the "real", initial NewEpochState, from the GenState
+
+genMockChainState ::
+  Reflect era =>
+  Proof era ->
+  GenState era ->
+  Gen (MockChainState era)
+genMockChainState proof gstate = pure $ MockChainState newepochstate (getSlot gstate) 0
+  where
+    ledgerstate = initialLedgerState gstate
+    newepochstate =
+      NewEpochState
+        { nesEL = EpochNo 0,
+          nesBprev = BlocksMade Map.empty,
+          nesBcur = BlocksMade Map.empty,
+          nesEs = makeEpochState gstate ledgerstate,
+          nesRu = SNothing,
+          nesPd = PoolDistr (gsInitialPoolDistr gstate),
+          stashedAVVMAddresses = stashedAVVMAddressesZero proof
+        }
+
+makeEpochState :: GenState era -> LedgerState era -> EpochState era
+makeEpochState gstate ledgerstate =
+  EpochState
+    { esAccountState = AccountState (getTreasury gstate) (getReserves gstate),
+      esSnapshots = def,
+      esLState = ledgerstate,
+      esPrevPp = gePParams (gsGenEnv gstate),
+      esPp = gePParams (gsGenEnv gstate),
+      esNonMyopic = def
+    }
+
+-- ==============================================================================
+
+-- =====================================================================
+-- HasTrace instance of MOCKCHAIN depends on STS(MOCKCHAIN era) instance
+-- We show the type family instances here for reference.
+{-
+instance STS (MOCKCHAIN era)
+  where
+  type State (MOCKCHAIN era) = MockChainState era
+  type Signal (MOCKCHAIN era) = (MockBlock era,UTxO era)
+  type Environment (MOCKCHAIN era) = ()
+-}
```

Comment:
Is this still needed?
```suggestion
```

Reply:
Yes I believe it is. One gets method types that are written in terms of associated type families of two unrelated classes (STS and HasTrace), it is very confusing, so having this guide makes it clear what types the methods of the HasTrace class should be.